# _Data Preparation and Cleaning_

__While the survey responses contain a lot of information, we are going to limit our analysis to the following areas:__

#### - Demographics
#### - Programming skills, experience, preferences, etc
#### - Employment-related information, preferences & opinions

Firstly, import the __Data_loading__ notebook to use here.

In [1]:
%run Data_loading.ipynb

__Let's select a subset of columns with relavant data for analysis.__

In [2]:
survey_raw_df.columns

Index(['ResponseId', 'MainBranch', 'Age', 'Employment', 'RemoteWork', 'Check',
       'CodingActivities', 'EdLevel', 'LearnCode', 'LearnCodeOnline',
       ...
       'JobSatPoints_6', 'JobSatPoints_7', 'JobSatPoints_8', 'JobSatPoints_9',
       'JobSatPoints_10', 'JobSatPoints_11', 'SurveyLength', 'SurveyEase',
       'ConvertedCompYearly', 'JobSat'],
      dtype='object', length=114)

In [3]:
survey_raw_df

ResponseId                      MainBranch                 Age  \
0               1  I am a developer by profession  Under 18 years old   
1               2  I am a developer by profession     35-44 years old   
2               3  I am a developer by profession     45-54 years old   
3               4           I am learning to code     18-24 years old   
4               5  I am a developer by profession     18-24 years old   
...           ...                             ...                 ...   
65432       65433  I am a developer by profession     18-24 years old   
65433       65434  I am a developer by profession     25-34 years old   
65434       65435  I am a developer by profession     25-34 years old   
65435       65436  I am a developer by profession     18-24 years old   
65436       65437     I code primarily as a hobby     18-24 years old   

                Employment                            RemoteWork   Check  \
0      Employed, full-time                                Remote  Apples   
1      Employed, full-time                                Remote  Apples   
2      Employed, full-time                                Remote  Apples   
3       Student, full-time                                   NaN  Apples   
4       Student, full-time                                   NaN  Apples   
...                    ...                                   ...     ...   
65432  Employed, full-time                                Remote  Apples   
65433  Employed, full-time                                Remote  Apples   
65434  Employed, full-time                             In-person  Apples   
65435  Employed, full-time  Hybrid (some remote, some in-person)  Apples   
65436   Student, full-time                                   NaN  Apples   

                                        CodingActivities  \
0                                                  Hobby   
1      Hobby;Contribute to open-source projects;Other...   
2      Hobby;Contribute to open-source projects;Other...   
3                                                    NaN   
4                                                    NaN   
...                                                  ...   
65432                      Hobby;School or academic work   
65433           Hobby;Contribute to open-source projects   
65434                                              Hobby   
65435  Hobby;Contribute to open-source projects;Profe...   
65436                                                NaN   

                                                 EdLevel  \
0                              Primary/elementary school   
1           Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
2        Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
3      Some college/university study without earning ...   
4      Secondary school (e.g. American high school, G...   
...                                                  ...   
65432       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
65433                                                NaN   
65434       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
65435  Secondary school (e.g. American high school, G...   
65436                                                NaN   

                                               LearnCode  \
0                                 Books / Physical media   
1      Books / Physical media;Colleague;On the job tr...   
2      Books / Physical media;Colleague;On the job tr...   
3      Other online resources (e.g., videos, blogs, f...   
4      Other online resources (e.g., videos, blogs, f...   
...                                                  ...   
65432  On the job training;School (i.e., University, ...   
65433                                                NaN   
65434  Other online resources (e.g., videos, blogs, f...   
65435  On the job training;Other online resources (e....   
65436                                                NaN   

                                         LearnCodeOnli

In [4]:
survey_selected_columns = [
    # Demograhpics
    'Country',
    'Age',
    'EdLevel',
    #Programming experience
    'LearnCode',
    'YearsCode',
    'YearsCodePro',
    'LanguageHaveWorkedWith',
    'WebframeHaveWorkedWith',
    'AIToolCurrently Using',
    #Employment related stuff
    'Employment',
    'RemoteWork',
    'DevType',
    'JobSat',
    'Industry',
]

schema_selected_columns = [
    # Demograhpics
    'Country',
    'Age',
    'EdLevel',
    #Programming experience
    'LearnCode',
    'YearsCode',
    'YearsCodePro',
    'Language',
    'Webframe',
    'AITool',
    #Employment related stuff
    'Employment',
    'RemoteWork',
    'DevType',
    'JobSat',
    'Industry',
]

In [5]:
len(survey_selected_columns)

18

In [6]:
len(schema_selected_columns)

15

Let's extract a copy of data from these columns into a new data frame __'survey_df'__, which we can continue to modify further without affecting the original data frame.

In [7]:
survey_df = survey_raw_df[survey_selected_columns].copy()

In [8]:
schema = schema_raw[schema_selected_columns]

Let's view some basic information about the data frame.

In [9]:
survey_df.shape

(65437, 18)

In [10]:
schema.shape

(15,)

In [11]:
survey_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 18 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Country                        58930 non-null  object 
 1   Age                            65437 non-null  object 
 2   EdLevel                        60784 non-null  object 
 3   LearnCode                      60488 non-null  object 
 4   YearsCode                      59869 non-null  object 
 5   YearsCodePro                   51610 non-null  object 
 6   LanguageHaveWorkedWith         59745 non-null  object 
 7   LanguageWantToWorkWith         55752 non-null  object 
 8   WebframeHaveWorkedWith         45161 non-null  object 
 9   WebframeWantToWorkWith         38535 non-null  object 
 10  AIToolCurrently Using          35072 non-null  object 
 11  AIToolInterested in Using      30691 non-null  object 
 12  AIToolNot interested in Using  24414 non-null 

Most columns have the data type __objects__, either because they contains values of different types, or they contain empty values, which are represented using __NaN__. It appears that some of the columns contains some empty values, since the Not-Null count for many columns is lower than the total number of rows(65437).

We'll have to deal with empty values and manually adjust the data type for each column on a case-by-case basis.

Only two one column has dtype as numeric column(JobSat), even though there are few other columns which have mostly numeric values. To make our analysis earsier, let's convert some other columns into numeric data types, while ignoring any non-numeric value.

In [12]:
schema.YearsCode    # questions

'Including any education, how many years have you been coding in total?'

In [13]:
survey_df.YearsCode.unique()    # unique answers

array([nan, '20', '37', '4', '9', '10', '7', '1', '15', '30', '31', '6',
       '12', '22', '5', '36', '25', '44', '24', '18', '3', '8',
       'More than 50 years', '11', '29', '40', '39', '2', '42', '34',
       '19', '35', '16', '33', '13', '23', '14', '28', '17', '21', '43',
       '46', '26', '32', '41', '45', '27', '38', '50', '48', '47',
       'Less than 1 year', '49'], dtype=object)

In [14]:
schema.YearsCodePro

'NOT including education, how many years have you coded professionally (as a part of your work)?'

In [15]:
survey_df.YearsCodePro.unique()

array([nan, '17', '27', '7', '11', '25', '12', '10', '3',
       'Less than 1 year', '18', '37', '15', '20', '6', '2', '16', '8',
       '14', '4', '45', '1', '24', '29', '5', '30', '26', '9', '33', '13',
       '35', '23', '22', '31', '19', '21', '28', '34', '32', '40', '50',
       '39', '44', '42', '41', '36', '38', 'More than 50 years', '43',
       '47', '48', '46', '49'], dtype=object)

In [16]:
survey_df['YearsCode'] = pd.to_numeric(survey_df.YearsCode, errors='coerce')
survey_df['YearsCodePro'] = pd.to_numeric(survey_df.YearsCodePro, errors='coerce')

Let's now see some __basic statistics__ about the numeric columns.

In [17]:
survey_df.describe()

YearsCode  YearsCodePro        JobSat
count  59046.000000  48704.000000  29126.000000
mean      14.175998     10.739652      6.935041
std       10.361047      8.935175      2.088259
min        1.000000      1.000000      0.000000
25%        6.000000      4.000000      6.000000
50%       11.000000      8.000000      7.000000
75%       20.000000     15.000000      8.000000
max       50.000000     50.000000     10.000000

These are only three numeric value columns and the data is almost cleaned and ready for the analysis. Let's look at the sample of rows from the data frame.

In [18]:
survey_df.sample(10)

Country                 Age  \
17922                 Australia     35-44 years old   
565    United States of America     35-44 years old   
51323                     Kenya     18-24 years old   
25246                     India  Under 18 years old   
62543                     Spain     35-44 years old   
14883               Netherlands  Under 18 years old   
20649                     India     25-34 years old   
45870               New Zealand     35-44 years old   
48468  United States of America     55-64 years old   
50614                 Argentina     35-44 years old   

                                                 EdLevel  \
17922       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
565                  Associate degree (A.A., A.S., etc.)   
51323  Some college/university study without earning ...   
25246  Secondary school (e.g. American high school, G...   
62543  Secondary school (e.g. American high school, G...   
14883  Secondary school (e.g. American high school, G...   
20649    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
45870     Professional degree (JD, MD, Ph.D, Ed.D, etc.)   
48468  Secondary school (e.g. American high school, G...   
50614  Some college/university study without earning ...   

                                               LearnCode  YearsCode  \
17922  Books / Physical media;On the job training;Oth...       24.0   
565    Books / Physical media;Other online resources ...       15.0   
51323                            Other (please specify):        1.0   
25246  Other online resources (e.g., videos, blogs, f...        NaN   
62543     Books / Physical media;Other (please specify):        9.0   
14883  Books / Physical media;Other online resources ...        6.0   
20649  On the job training;Other online resources (e....        5.0   
45870  Other online resources (e.g., videos, blogs, f...        6.0   
48468  On the job training;Other online resources (e....       45.0   
50614  Books / Physical media;Other online resources ...       15.0   

       YearsCodePro                             LanguageHaveWorkedWith  \
17922          20.0                       Bash/Shell (all shells);Java   
565            15.0  C#;HTML/CSS;PowerShell;Python;SQL;VBA;Visual B...   
51323           NaN                                      HTML/CSS;Java   
25246           NaN        C;C++;HTML/CSS;JavaScript;MATLAB;Python;SQL   
62543           7.0  Bash/Shell (all shells);Go;HTML/CSS;JavaScript...   
14883           NaN  Bash/Shell (all shells);HTML/CSS;JavaScript;Ko...   
20649           3.0                           HTML/CSS;Java;JavaScript   
45870           6.0                                              R;SQL   
48468          26.0              C#;HTML/CSS;JavaScript;PowerShell;SQL   
50614           5.0  Bash/Shell (all shells);Elixir;Go;HTML/CSS;Jav...   

                                  LanguageWantToWorkWith  \
17922                              Assembly;C++;GDScript   
565                               C#;Visual Basic (.Net)   
51323                         JavaScript;Kotlin;PHP;Rust   
25246  Bash/Shell (all shells);C;C++;Elixir;Go;Java;J...   
62543                Bash/Shell (all shells);Go;Rust;SQL   
14883  Bash/Shell (all shells);C;Haskell;HTML/CSS;Pyt...   
20649                                             Kotlin   
45870                                              R;SQL   
48468              C#;HTML/CSS;JavaScript;PowerShell;SQL   
50614  Bash/Shell (all shells);Elixir;HTML/CSS;JavaSc...   

                   WebframeHaveWorkedWith  \
17922                                 NaN   
565                                Blazor   
51323                               Yii 2   
25246                              jQuery   
62543      Laravel;Node.js;Symfony;Vue.js   
14883                        Flask;jQuery   
20649                  jQuery;Spring Boot   
45870                                 NaN   
48468  ASP.NET;ASP.NET CORE;Blazor;jQuery   
50614                             Phoenix   

   

### Data preparation and cleaning is complete now the data is ready for the visualization and analysis.